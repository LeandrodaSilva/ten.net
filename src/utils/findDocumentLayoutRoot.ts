import documentLayout from "../assets/document.html" with { type: "text" };

/**
 * Finds and returns the content of a document layout root file.
 *
 * Attempts to read a `document.html` file from the specified application path.
 * If the file doesn't exist or cannot be read, returns a default HTML template
 * with a placeholder for content injection.
 *
 * @param appPath - The base path of the application where the document.html file should be located
 * @returns The content of the document.html file, or a default HTML template string if the file is not found
 *
 * @example
 * ```typescript
 * const layout = findDocumentLayoutRoot('/path/to/app');
 * // Returns content of /path/to/app/document.html or default template
 * ```
 */
export function findDocumentLayoutRoot(appPath?: string): string {
  if (appPath) {
	  const rootLayoutPath = `${appPath}/document.html`;
	  try {
		  Deno.lstatSync(rootLayoutPath);
		  return Deno.readTextFileSync(rootLayoutPath);
	  } catch {}
  }
	return <string>documentLayout;
}
