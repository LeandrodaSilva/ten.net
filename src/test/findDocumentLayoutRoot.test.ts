import { assertEquals } from "@deno-assert";
import { findDocumentLayoutRoot } from "../utils/findDocumentLayoutRoot.ts";
const TEST_DIR = "./test_temp";
const TEST_APP_PATH = `${TEST_DIR}/app`;
const DOCUMENT_HTML_PATH = `${TEST_APP_PATH}/document.html`;
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("findDocumentLayoutRoot - returns file content when document.html exists", async (t) => {
  // Setup
  Deno.mkdirSync(TEST_APP_PATH, { recursive: true });
  Deno.writeTextFileSync(DOCUMENT_HTML_PATH, t.toString());

  try {
    // Act
    const result = findDocumentLayoutRoot(TEST_APP_PATH);

    // Assert
    await assertSnapshot(t, result);
  } finally {
    // Cleanup
    Deno.removeSync(TEST_DIR, { recursive: true });
  }
});

Deno.test("findDocumentLayoutRoot - returns default template when document.html does not exist", async (t) => {
  // Setup
  const nonExistentPath = "./non_existent_path";

  // Act
  const result = findDocumentLayoutRoot(nonExistentPath);

  // Assert
  await assertSnapshot(t, result);
});

Deno.test("findDocumentLayoutRoot - returns default template when file exists but cannot be read", async (t) => {
  // Setup
  Deno.mkdirSync(TEST_APP_PATH, { recursive: true });
  Deno.writeTextFileSync(DOCUMENT_HTML_PATH, "test content");

  // Make file unreadable by changing permissions (Unix-like systems)
  try {
    Deno.chmodSync(DOCUMENT_HTML_PATH, 0o000);
  } catch {
    // Skip this test on systems that don't support chmod
    Deno.removeSync(TEST_DIR, { recursive: true });
    return;
  }

  try {
    // Act
    const result = findDocumentLayoutRoot(TEST_APP_PATH);

    // Assert
    await assertSnapshot(t, result);
  } finally {
    // Cleanup
    try {
      Deno.chmodSync(DOCUMENT_HTML_PATH, 0o644);
    } catch {
      // Ignore cleanup errors
    }
    Deno.removeSync(TEST_DIR, { recursive: true });
  }
});

Deno.test("findDocumentLayoutRoot - handles empty file", () => {
  // Setup
  Deno.mkdirSync(TEST_APP_PATH, { recursive: true });
  Deno.writeTextFileSync(DOCUMENT_HTML_PATH, "");

  try {
    // Act
    const result = findDocumentLayoutRoot(TEST_APP_PATH);

    // Assert
    assertEquals(result, "");
  } finally {
    // Cleanup
    Deno.removeSync(TEST_DIR, { recursive: true });
  }
});

Deno.test("findDocumentLayoutRoot - handles different app paths", () => {
  // Setup
  const customAppPath = `${TEST_DIR}/custom/app/path`;
  const customDocumentPath = `${customAppPath}/document.html`;
  Deno.mkdirSync(customAppPath, { recursive: true });
  const expectedContent = "<html><body>Custom path content</body></html>";
  Deno.writeTextFileSync(customDocumentPath, expectedContent);

  try {
    // Act
    const result = findDocumentLayoutRoot(customAppPath);

    // Assert
    assertEquals(result, expectedContent);
  } finally {
    // Cleanup
    Deno.removeSync(TEST_DIR, { recursive: true });
  }
});
