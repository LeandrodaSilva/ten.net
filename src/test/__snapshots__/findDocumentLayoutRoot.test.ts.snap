export const snapshot = {};

snapshot[`findDocumentLayoutRoot - returns file content when document.html exists 1`] = `"[object TestContext]"`;

snapshot[`findDocumentLayoutRoot - returns default template when document.html does not exist 1`] = `
'<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="https://cdn.jsdelivr.net/npm/@tailwindcss/browser@4">
    </script>
    <script
      src="https://cdn.jsdelivr.net/npm/@tailwindplus/elements@1"
      type="module"
    >
    </script>
    <title>Ten.net</title>
  </head>
  <body>{{content}}</body>
</html>
'
`;

snapshot[`findDocumentLayoutRoot - returns default template when file exists but cannot be read 1`] = `
'<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="https://cdn.jsdelivr.net/npm/@tailwindcss/browser@4">
    </script>
    <script
      src="https://cdn.jsdelivr.net/npm/@tailwindplus/elements@1"
      type="module"
    >
    </script>
    <title>Ten.net</title>
  </head>
  <body>{{content}}</body>
</html>
'
`;
